## Import some dependencies

In [1]:
# Set the seed value for the notebook so the results are reproducible
import numpy as np
from numpy.random import seed
seed(1)
#wanted to take a look
np.random.rand(3)

array([4.17022005e-01, 7.20324493e-01, 1.14374817e-04])

In [2]:
import pandas as pd
import os

In [3]:
import tensorflow
tensorflow.keras.__version__

'2.2.4-tf'

## Read in and clean up data

In [4]:
test_data_orig = pd.read_csv(os.path.join('data', 'loanDataTest.csv'))
pd.set_option('display.max_columns', None)
test_data_orig

Loan_ID Gender Married Dependents     Education Self_Employed  \
0    LP001015   Male     Yes          0      Graduate            No   
1    LP001022   Male     Yes          1      Graduate            No   
2    LP001031   Male     Yes          2      Graduate            No   
3    LP001035   Male     Yes          2      Graduate            No   
4    LP001051   Male      No          0  Not Graduate            No   
..        ...    ...     ...        ...           ...           ...   
362  LP002971   Male     Yes         3+  Not Graduate           Yes   
363  LP002975   Male     Yes          0      Graduate            No   
364  LP002980   Male      No          0      Graduate            No   
365  LP002986   Male     Yes          0      Graduate            No   
366  LP002989   Male      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5720                  0       110.0             360.0   
1               3076               1500       126.0             360.0   
2               5000               1800       208.0             360.0   
3               2340               2546       100.0             360.0   
4               3276                  0        78.0             360.0   
..               ...                ...         ...               ...   
362             4009               1777       113.0             360.0   
363             4158                709       115.0             360.0   
364             3250               1993       126.0             360.0   
365             5000               2393       158.0             360.0   
366             9200                  0        98.0             180.0   

     Credit_History Property_Area  
0               1.0         Urban  
1               1.0         Urban  
2               1.0         Urban  
3               NaN         Urban  
4               1.0         Urban  
..              ...           ...  
362             1.0         Urban  
363             1.0         Urban  
364             NaN     Semiurban  
365             1.0         Rural  
366             1.0         Rural  

[367 rows x 12 columns]

In [5]:
train_data_orig = pd.read_csv(os.path.join('data', 'loanDataTrain.csv'))
pd.set_option('display.max_columns', None)
train_data_orig

Loan_ID  Gender Married Dependents     Education Self_Employed  \
0    LP001002    Male      No          0      Graduate            No   
1    LP001003    Male     Yes          1      Graduate            No   
2    LP001005    Male     Yes          0      Graduate           Yes   
3    LP001006    Male     Yes          0  Not Graduate            No   
4    LP001008    Male      No          0      Graduate            No   
..        ...     ...     ...        ...           ...           ...   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849                0.0         NaN             360.0   
1               4583             1508.0       128.0             360.0   
2               3000                0.0        66.0             360.0   
3               2583             2358.0       120.0             360.0   
4               6000                0.0       141.0             360.0   
..               ...                ...         ...               ...   
609             2900                0.0        71.0             360.0   
610             4106                0.0        40.0             180.0   
611             8072              240.0       253.0             360.0   
612             7583                0.0       187.0             360.0   
613             4583                0.0       133.0             360.0   

     Credit_History Property_Area Loan_Status  
0               1.0         Urban           Y  
1               1.0         Rural           N  
2               1.0         Urban           Y  
3               1.0         Urban           Y  
4               1.0         Urban           Y  
..              ...           ...         ...  
609             1.0         Rural           Y  
610             1.0         Rural           Y  
611             1.0         Urban           Y  
612             1.0         Urban           Y  
613             0.0     Semiurban           N  

[614 rows x 13 columns]

In [6]:
#Check values in test and train sets
test_Gender = test_data_orig.Gender.unique()
train_Gender = train_data_orig.Gender.unique()
test_Dependents = test_data_orig.Dependents.unique()
train_Dependents = train_data_orig.Dependents.unique()
test_Education = test_data_orig.Education.unique()
train_Education = train_data_orig.Education.unique()
test_Self_Employed = test_data_orig.Self_Employed.unique()
train_Self_Employed = train_data_orig.Self_Employed.unique()
test_ApplicantIncome = test_data_orig.ApplicantIncome.unique()
train_ApplicantIncome = train_data_orig.ApplicantIncome.unique()
test_CoapplicantIncome = test_data_orig.CoapplicantIncome.unique()
train_CoapplicantIncome = train_data_orig.CoapplicantIncome.unique()
test_LoanAmount = test_data_orig.LoanAmount.unique()
train_LoanAmount = train_data_orig.LoanAmount.unique()
test_Loan_Amount_Term = test_data_orig.Loan_Amount_Term.unique()
train_Loan_Amount_Term = train_data_orig.Loan_Amount_Term.unique()
test_Credit_History = test_data_orig.Credit_History.unique()
train_Credit_History = train_data_orig.Credit_History.unique()
test_Property_Area = test_data_orig.Property_Area.unique()
train_Property_Area = train_data_orig.Property_Area.unique()

train_Loan_Status = train_data_orig.Loan_Status.unique()

#Print out values
#test_Gender, train_Gender, test_Dependents, train_Dependents, test_Education, train_Education, test_Self_Employed, train_Self_Employed, test_ApplicantIncome, train_ApplicantIncome, test_CoapplicantIncome, train_CoapplicantIncome, test_LoanAmount, train_LoanAmount, test_Loan_Amount_Term, train_Loan_Amount_Term, test_Credit_History, train_Credit_History, test_Property_Area, train_Property_Area, train_Loan_Status

In [7]:
#drop nans
test_data = test_data_orig.dropna(axis=0)
train_data_pre_dummy = train_data_orig.dropna(axis=0)

test_Gender = test_data.Gender.unique()
train_Gender = train_data_pre_dummy.Gender.unique()
test_Dependents = test_data.Dependents.unique()
train_Dependents = train_data_pre_dummy.Dependents.unique()
test_Education = test_data.Education.unique()
train_Education = train_data_pre_dummy.Education.unique()
test_Self_Employed = test_data.Self_Employed.unique()
train_Self_Employed = train_data_pre_dummy.Self_Employed.unique()
test_ApplicantIncome = test_data.ApplicantIncome.unique()
train_ApplicantIncome = train_data_pre_dummy.ApplicantIncome.unique()
test_CoapplicantIncome = test_data.CoapplicantIncome.unique()
train_CoapplicantIncome = train_data_pre_dummy.CoapplicantIncome.unique()
test_LoanAmount = test_data.LoanAmount.unique()
train_LoanAmount = train_data_pre_dummy.LoanAmount.unique()
test_Loan_Amount_Term = test_data.Loan_Amount_Term.unique()
train_Loan_Amount_Term = train_data_pre_dummy.Loan_Amount_Term.unique()
test_Credit_History = test_data.Credit_History.unique()
train_Credit_History = train_data_pre_dummy.Credit_History.unique()
test_Property_Area = test_data.Property_Area.unique()
train_Property_Area = train_data_pre_dummy.Property_Area.unique()

train_Loan_Status = train_data_pre_dummy.Loan_Status.unique()

#Print out categorical values
test_Gender, train_Gender, test_Dependents, train_Dependents, test_Education, train_Education, test_Self_Employed, train_Self_Employed, train_Loan_Amount_Term, test_Credit_History, train_Credit_History, test_Property_Area, train_Property_Area, train_Loan_Status

(array(['Male', 'Female'], dtype=object),
 array(['Male', 'Female'], dtype=object),
 array(['0', '1', '2', '3+'], dtype=object),
 array(['1', '0', '2', '3+'], dtype=object),
 array(['Graduate', 'Not Graduate'], dtype=object),
 array(['Graduate', 'Not Graduate'], dtype=object),
 array(['No', 'Yes'], dtype=object),
 array(['No', 'Yes'], dtype=object),
 array([360., 120., 180.,  60., 300., 480., 240.,  36.,  84.]),
 array([1., 0.]),
 array([1., 0.]),
 array(['Urban', 'Semiurban', 'Rural'], dtype=object),
 array(['Rural', 'Urban', 'Semiurban'], dtype=object),
 array(['N', 'Y'], dtype=object))

In [8]:
#Make loan status a dummy variable, drop the "N" column
train_data = pd.get_dummies(train_data_pre_dummy, prefix=['Loan_Status'], columns=['Loan_Status'])
train_data.drop('Loan_Status_N', axis=1, inplace=True)
train_data

Loan_ID  Gender Married Dependents     Education Self_Employed  \
1    LP001003    Male     Yes          1      Graduate            No   
2    LP001005    Male     Yes          0      Graduate           Yes   
3    LP001006    Male     Yes          0  Not Graduate            No   
4    LP001008    Male      No          0      Graduate            No   
5    LP001011    Male     Yes          2      Graduate           Yes   
..        ...     ...     ...        ...           ...           ...   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1               4583             1508.0       128.0             360.0   
2               3000                0.0        66.0             360.0   
3               2583             2358.0       120.0             360.0   
4               6000                0.0       141.0             360.0   
5               5417             4196.0       267.0             360.0   
..               ...                ...         ...               ...   
609             2900                0.0        71.0             360.0   
610             4106                0.0        40.0             180.0   
611             8072              240.0       253.0             360.0   
612             7583                0.0       187.0             360.0   
613             4583                0.0       133.0             360.0   

     Credit_History Property_Area  Loan_Status_Y  
1               1.0         Rural              0  
2               1.0         Urban              1  
3               1.0         Urban              1  
4               1.0         Urban              1  
5               1.0         Urban              1  
..              ...           ...            ...  
609             1.0         Rural              1  
610             1.0         Rural              1  
611             1.0         Urban              1  
612             1.0         Urban              1  
613             0.0     Semiurban              0  

[480 rows x 13 columns]

In [9]:
#More dummy variables
train_data_more_dummies = pd.get_dummies(train_data, prefix=['Gender'], columns=['Gender'])
train_data_more_dummies.drop('Gender_Male', axis=1, inplace=True)
test_data_more_dummies = pd.get_dummies(test_data, prefix=['Gender'], columns=['Gender'])
test_data_more_dummies.drop('Gender_Male', axis=1, inplace=True)
train_data_more_dummies = pd.get_dummies(train_data_more_dummies, prefix=['Education'], columns=['Education'])
train_data_more_dummies.drop('Education_Not Graduate', axis=1, inplace=True)
test_data_more_dummies = pd.get_dummies(test_data_more_dummies, prefix=['Education'], columns=['Education'])
test_data_more_dummies.drop('Education_Not Graduate', axis=1, inplace=True)
train_data_more_dummies = pd.get_dummies(train_data_more_dummies, prefix=['Married'], columns=['Married'])
train_data_more_dummies.drop('Married_No', axis=1, inplace=True)
test_data_more_dummies = pd.get_dummies(test_data_more_dummies, prefix=['Married'], columns=['Married'])
test_data_more_dummies.drop('Married_No', axis=1, inplace=True)
train_data_more_dummies = pd.get_dummies(train_data_more_dummies, prefix=['Self_Employed'], columns=['Self_Employed'])
train_data_more_dummies.drop('Self_Employed_No', axis=1, inplace=True)
test_data_more_dummies = pd.get_dummies(test_data_more_dummies, prefix=['Self_Employed'], columns=['Self_Employed'])
test_data_more_dummies.drop('Self_Employed_No', axis=1, inplace=True)
train_data_more_dummies = pd.get_dummies(train_data_more_dummies, prefix=['Credit_History'], columns=['Credit_History'])
train_data_more_dummies.drop('Credit_History_0.0', axis=1, inplace=True)
test_data_more_dummies = pd.get_dummies(test_data_more_dummies, prefix=['Credit_History'], columns=['Credit_History'])
test_data_more_dummies.drop('Credit_History_0.0', axis=1, inplace=True)

train_data_more_dummies = pd.get_dummies(train_data_more_dummies, prefix=['Property_Area'], columns=['Property_Area'])
test_data_more_dummies = pd.get_dummies(test_data_more_dummies, prefix=['Property_Area'], columns=['Property_Area'])

train_data_more_dummies = pd.get_dummies(train_data_more_dummies, prefix=['Dependents'], columns=['Dependents'])
test_data_more_dummies = pd.get_dummies(test_data_more_dummies, prefix=['Dependents'], columns=['Dependents'])

train_data_more_dummies = pd.get_dummies(train_data_more_dummies, prefix=['Loan_Amount_Term'], columns=['Loan_Amount_Term'])
test_data_more_dummies = pd.get_dummies(test_data_more_dummies, prefix=['Loan_Amount_Term'], columns=['Loan_Amount_Term'])

train_data_more_dummies

Loan_ID  ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Status_Y  \
1    LP001003             4583             1508.0       128.0              0   
2    LP001005             3000                0.0        66.0              1   
3    LP001006             2583             2358.0       120.0              1   
4    LP001008             6000                0.0       141.0              1   
5    LP001011             5417             4196.0       267.0              1   
..        ...              ...                ...         ...            ...   
609  LP002978             2900                0.0        71.0              1   
610  LP002979             4106                0.0        40.0              1   
611  LP002983             8072              240.0       253.0              1   
612  LP002984             7583                0.0       187.0              1   
613  LP002990             4583                0.0       133.0              0   

     Gender_Female  Education_Graduate  Married_Yes  Self_Employed_Yes  \
1                0                   1            1                  0   
2                0                   1            1                  1   
3                0                   0            1                  0   
4                0                   1            0                  0   
5                0                   1            1                  1   
..             ...                 ...          ...                ...   
609              1                   1            0                  0   
610              0                   1            1                  0   
611              0                   1            1                  0   
612              0                   1            1                  0   
613              1                   1            0                  1   

     Credit_History_1.0  Property_Area_Rural  Property_Area_Semiurban  \
1                     1                    1                        0   
2                     1                    0                        0   
3                     1                    0                        0   
4                     1                    0                        0   
5                     1                    0                        0   
..                  ...                  ...                      ...   
609                   1                    1                        0   
610                   1                    1                        0   
611                   1                    0                        0   
612                   1                    0                        0   
613                   0                    0                        1   

     Property_Area_Urban  Dependents_0  Dependents_1  Dependents_2  \
1                      0             0             1             0   
2                      1             1             0             0   
3                      1             1             0             0   
4                      1             1             0             0   
5                      1             0             0             1   
..                   ...           ...           ...           ...   
609                    0             1             0             0   
610                    0             0             0             0   
611                    1             0             1             0   
612                    1             0             0             1   
613                    0             1             0             0   

     Dependents_3+  Loan_Amount_Term_36.0  Loan_Amount_Term_60.0  \
1                0                      0                      0   
2                0                      0                      0   
3                0                      0                      0   
4                0                      0                      0   
5                0                      0                      0   
..             ...            

## First run: all columns from above

In [14]:
X = train_data_more_dummies.drop(["Loan_Status_Y", "Loan_ID"], axis=1)
y = train_data_more_dummies["Loan_Status_Y"]
print(X.shape, y.shape)

(480, 24) (480,)


In [15]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, MinMaxScaler
from tensorflow.keras.utils import to_categorical

In [16]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, random_state=1)

In [17]:
X_scaler = MinMaxScaler().fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [18]:
# Step 1: Label-encode data set
label_encoder = LabelEncoder()
label_encoder.fit(y_train)
encoded_y_train = label_encoder.transform(y_train)
encoded_y_test = label_encoder.transform(y_test)

In [19]:
# Step 2: Convert encoded labels to one-hot-encoding
y_train_categorical = to_categorical(encoded_y_train)
y_test_categorical = to_categorical(encoded_y_test)

In [20]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [25]:
# Create model and add layers
model = Sequential()
model.add(Dense(units=100, activation='relu', input_dim=24))
model.add(Dense(units=100, activation='relu'))
model.add(Dense(units=2, activation='softmax'))

In [26]:
# Compile and fit the model
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

In [27]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 100)               2500      
_________________________________________________________________
dense_4 (Dense)              (None, 100)               10100     
_________________________________________________________________
dense_5 (Dense)              (None, 2)                 202       
Total params: 12,802
Trainable params: 12,802
Non-trainable params: 0
_________________________________________________________________


In [28]:
model.fit(
    X_train_scaled,
    y_train_categorical,
    epochs=60,
    shuffle=True,
    verbose=2)

Train on 360 samples
Epoch 1/60
360/360 - 4s - loss: 0.5819 - accuracy: 0.7083
Epoch 2/60
360/360 - 0s - loss: 0.5260 - accuracy: 0.7250
Epoch 3/60
360/360 - 0s - loss: 0.4924 - accuracy: 0.7861
Epoch 4/60
360/360 - 0s - loss: 0.4627 - accuracy: 0.8056
Epoch 5/60
360/360 - 0s - loss: 0.4368 - accuracy: 0.8278
Epoch 6/60
360/360 - 0s - loss: 0.4218 - accuracy: 0.8250
Epoch 7/60
360/360 - 0s - loss: 0.4082 - accuracy: 0.8250
Epoch 8/60
360/360 - 0s - loss: 0.3999 - accuracy: 0.8389
Epoch 9/60
360/360 - 0s - loss: 0.4016 - accuracy: 0.8306
Epoch 10/60
360/360 - 0s - loss: 0.3843 - accuracy: 0.8389
Epoch 11/60
360/360 - 0s - loss: 0.3762 - accuracy: 0.8389
Epoch 12/60
360/360 - 0s - loss: 0.3715 - accuracy: 0.8472
Epoch 13/60
360/360 - 0s - loss: 0.3653 - accuracy: 0.8500
Epoch 14/60
360/360 - 0s - loss: 0.3607 - accuracy: 0.8500
Epoch 15/60
360/360 - 0s - loss: 0.3563 - accuracy: 0.8500
Epoch 16/60
360/360 - 0s - loss: 0.3665 - accuracy: 0.8639
Epoch 17/60
360/360 - 0s - loss: 0.3587 - ac

In [29]:
model_loss, model_accuracy = model.evaluate(
    X_test_scaled, y_test_categorical, verbose=2)
print(
    f"Normal Neural Network 1 - Loss: {model_loss}, Accuracy: {model_accuracy}")

120/1 - 0s - loss: 0.8075 - accuracy: 0.7000
Normal Neural Network 1 - Loss: 0.9098991235097249, Accuracy: 0.699999988079071


First run results

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
dense_3 (Dense)              (None, 100)               2500      
_________________________________________________________________
dense_4 (Dense)              (None, 100)               10100     
_________________________________________________________________
dense_5 (Dense)              (None, 2)                 202       
=================================================================
Total params: 12,802
Trainable params: 12,802
Non-trainable params: 0
    
120/1 - 0s - loss: 0.8075 - accuracy: 0.7000
Normal Neural Network 1 - Loss: 0.9098991235097249, Accuracy: 0.699999988079071

## Second run: take out Loan_Amount_Term columns other than "480": accuracy a bit higher

In [30]:
X = train_data_more_dummies.drop(["Loan_Status_Y", "Loan_ID", 'Loan_Amount_Term_360.0', 'Loan_Amount_Term_180.0','Loan_Amount_Term_300.0', 'Loan_Amount_Term_240.0', 'Loan_Amount_Term_60.0', 'Loan_Amount_Term_120.0','Loan_Amount_Term_84.0', 'Loan_Amount_Term_36.0'], axis=1)
y = train_data_more_dummies["Loan_Status_Y"]
print(X.shape, y.shape)

(480, 16) (480,)


In [31]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, MinMaxScaler
from tensorflow.keras.utils import to_categorical

In [32]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, random_state=1)

In [33]:
X_scaler = MinMaxScaler().fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [34]:
# Step 1: Label-encode data set
label_encoder = LabelEncoder()
label_encoder.fit(y_train)
encoded_y_train = label_encoder.transform(y_train)
encoded_y_test = label_encoder.transform(y_test)

In [35]:
# Step 2: Convert encoded labels to one-hot-encoding
y_train_categorical = to_categorical(encoded_y_train)
y_test_categorical = to_categorical(encoded_y_test)

In [36]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [37]:
# Create model and add layers
model = Sequential()
model.add(Dense(units=100, activation='relu', input_dim=16))
model.add(Dense(units=100, activation='relu'))
model.add(Dense(units=2, activation='softmax'))

In [38]:
# Compile and fit the model
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

In [39]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_6 (Dense)              (None, 100)               1700      
_________________________________________________________________
dense_7 (Dense)              (None, 100)               10100     
_________________________________________________________________
dense_8 (Dense)              (None, 2)                 202       
Total params: 12,002
Trainable params: 12,002
Non-trainable params: 0
_________________________________________________________________


In [40]:
model.fit(
    X_train_scaled,
    y_train_categorical,
    epochs=60,
    shuffle=True,
    verbose=2)

Train on 360 samples
Epoch 1/60
360/360 - 1s - loss: 0.6256 - accuracy: 0.6528
Epoch 2/60
360/360 - 0s - loss: 0.5603 - accuracy: 0.7083
Epoch 3/60
360/360 - 0s - loss: 0.5156 - accuracy: 0.7500
Epoch 4/60
360/360 - 0s - loss: 0.4859 - accuracy: 0.8000
Epoch 5/60
360/360 - 0s - loss: 0.4561 - accuracy: 0.8194
Epoch 6/60
360/360 - 0s - loss: 0.4368 - accuracy: 0.8222
Epoch 7/60
360/360 - 0s - loss: 0.4250 - accuracy: 0.8278
Epoch 8/60
360/360 - 0s - loss: 0.4136 - accuracy: 0.8306
Epoch 9/60
360/360 - 0s - loss: 0.4046 - accuracy: 0.8361
Epoch 10/60
360/360 - 0s - loss: 0.3980 - accuracy: 0.8417
Epoch 11/60
360/360 - 0s - loss: 0.3939 - accuracy: 0.8389
Epoch 12/60
360/360 - 0s - loss: 0.3881 - accuracy: 0.8333
Epoch 13/60
360/360 - 0s - loss: 0.3850 - accuracy: 0.8444
Epoch 14/60
360/360 - 0s - loss: 0.3836 - accuracy: 0.8389
Epoch 15/60
360/360 - 0s - loss: 0.3791 - accuracy: 0.8389
Epoch 16/60
360/360 - 0s - loss: 0.3787 - accuracy: 0.8444
Epoch 17/60
360/360 - 0s - loss: 0.3750 - ac

In [41]:
model_loss, model_accuracy = model.evaluate(
    X_test_scaled, y_test_categorical, verbose=2)
print(
    f"Normal Neural Network 1 - Loss: {model_loss}, Accuracy: {model_accuracy}")

120/1 - 0s - loss: 0.7736 - accuracy: 0.7333
Normal Neural Network 1 - Loss: 0.7851928194363912, Accuracy: 0.7333333492279053


Run 2 summary

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
dense_6 (Dense)              (None, 100)               1700      
_________________________________________________________________
dense_7 (Dense)              (None, 100)               10100     
_________________________________________________________________
dense_8 (Dense)              (None, 2)                 202       
=================================================================
Total params: 12,002
Trainable params: 12,002
Non-trainable params: 0
    
120/1 - 0s - loss: 0.7736 - accuracy: 0.7333
Normal Neural Network 1 - Loss: 0.7851928194363912, Accuracy: 0.7333333492279053


## Third run: take out Loan_Amount_Term and LoanAmount columns, since these could be considered results: little change in accuracy

In [43]:
X = train_data_more_dummies.drop(["Loan_Status_Y", "Loan_ID", 'LoanAmount', 'Loan_Amount_Term_480.0', 'Loan_Amount_Term_360.0', 'Loan_Amount_Term_180.0', 'Loan_Amount_Term_300.0', 'Loan_Amount_Term_240.0', 'Loan_Amount_Term_60.0', 'Loan_Amount_Term_120.0', 'Loan_Amount_Term_84.0', 'Loan_Amount_Term_36.0'], axis=1)
y = train_data_more_dummies["Loan_Status_Y"]
print(X.shape, y.shape)

(480, 14) (480,)


In [44]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, MinMaxScaler
from tensorflow.keras.utils import to_categorical

In [45]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, random_state=1)

In [46]:
X_scaler = MinMaxScaler().fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [47]:
# Step 1: Label-encode data set
label_encoder = LabelEncoder()
label_encoder.fit(y_train)
encoded_y_train = label_encoder.transform(y_train)
encoded_y_test = label_encoder.transform(y_test)

In [48]:
# Step 2: Convert encoded labels to one-hot-encoding
y_train_categorical = to_categorical(encoded_y_train)
y_test_categorical = to_categorical(encoded_y_test)

In [49]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [50]:
# Create model and add layers
model = Sequential()
model.add(Dense(units=100, activation='relu', input_dim=14))
model.add(Dense(units=100, activation='relu'))
model.add(Dense(units=2, activation='softmax'))

In [51]:
# Compile and fit the model
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

In [52]:
model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_9 (Dense)              (None, 100)               1500      
_________________________________________________________________
dense_10 (Dense)             (None, 100)               10100     
_________________________________________________________________
dense_11 (Dense)             (None, 2)                 202       
Total params: 11,802
Trainable params: 11,802
Non-trainable params: 0
_________________________________________________________________


In [53]:
model.fit(
    X_train_scaled,
    y_train_categorical,
    epochs=60,
    shuffle=True,
    verbose=2)

Train on 360 samples
Epoch 1/60
360/360 - 1s - loss: 0.5876 - accuracy: 0.7056
Epoch 2/60
360/360 - 0s - loss: 0.5426 - accuracy: 0.7222
Epoch 3/60
360/360 - 0s - loss: 0.5168 - accuracy: 0.7333
Epoch 4/60
360/360 - 0s - loss: 0.4953 - accuracy: 0.7861
Epoch 5/60
360/360 - 0s - loss: 0.4776 - accuracy: 0.8056
Epoch 6/60
360/360 - 0s - loss: 0.4578 - accuracy: 0.8139
Epoch 7/60
360/360 - 0s - loss: 0.4453 - accuracy: 0.8222
Epoch 8/60
360/360 - 0s - loss: 0.4348 - accuracy: 0.8250
Epoch 9/60
360/360 - 0s - loss: 0.4221 - accuracy: 0.8278
Epoch 10/60
360/360 - 0s - loss: 0.4183 - accuracy: 0.8333
Epoch 11/60
360/360 - 0s - loss: 0.4148 - accuracy: 0.8306
Epoch 12/60
360/360 - 0s - loss: 0.4074 - accuracy: 0.8306
Epoch 13/60
360/360 - 0s - loss: 0.4067 - accuracy: 0.8306
Epoch 14/60
360/360 - 0s - loss: 0.4139 - accuracy: 0.8389
Epoch 15/60
360/360 - 0s - loss: 0.4032 - accuracy: 0.8306
Epoch 16/60
360/360 - 0s - loss: 0.3982 - accuracy: 0.8306
Epoch 17/60
360/360 - 0s - loss: 0.3943 - ac

In [54]:
model_loss, model_accuracy = model.evaluate(
    X_test_scaled, y_test_categorical, verbose=2)
print(
    f"Normal Neural Network 1 - Loss: {model_loss}, Accuracy: {model_accuracy}")

120/1 - 0s - loss: 0.7815 - accuracy: 0.7333
Normal Neural Network 1 - Loss: 0.8348241329193116, Accuracy: 0.7333333492279053


In [ ]:
Run 3 summary

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
dense_9 (Dense)              (None, 100)               1500      
_________________________________________________________________
dense_10 (Dense)             (None, 100)               10100     
_________________________________________________________________
dense_11 (Dense)             (None, 2)                 202       
=================================================================
Total params: 11,802
Trainable params: 11,802
Non-trainable params: 0
    
120/1 - 0s - loss: 0.7815 - accuracy: 0.7333
Normal Neural Network 1 - Loss: 0.8348241329193116, Accuracy: 0.7333333492279053